### PR TITLE
open and read timeout

### PR DESCRIPTION
**Pull request by rodrhern**
AVIOInterrupt which allows to define open and read timeuot

### CODE REVIEW

**Comment by mikeboers** [comment]
This is cool! I'll merge this soon.

**Comment by rodrhern** [comment]
great!! feel free to change the code if you need

**Comment by gemfield** [comment]
This is really what I needed, I have met the av.open() blocked by some rtmp streams...

**Comment by mikeboers** [comment]
I still want to pull this in...

**Comment by rodrhern** [comment]
That would be nice. I can rebase the code.

Do you think that the dictionary I made is the best way of passing the options to the code??. 
This options are not supposed to be forwarded to ffmpeg, they are removed in the wrapper, but I'm using the options kwarg and poping the timeout info.

**Comment by angelfw** [comment]
Any update for this PR? This feature is quite necessary as we need a clean quit on some unreachable rtsp streams.

**Comment by hugeturing** [comment]
I desperately need this feature. Can someone help to resolve this PR? Thanks.

**Comment by tianhuyang** [comment]
I desperately need this feature. Can someone help to resolve this PR? Thanks.

**Comment by jlaine** [comment]
Any chance of fixing the compilation error please?

**Comment by jlaine** [comment]
The build failure on Windows is still exactly the same (see the AppVeyor build):

`error C2079: 'start_time' uses undefined struct 'timeval'`

**Comment by rodrhern** [comment]
Thanks, this will take me longer, since I don't regularly use windows, and I have to find a spare time to do it

**Comment by jlaine** [comment]
This might help, try using <sys/time.h>:

https://github.com/cython/cython/blob/master/Cython/Includes/posix/time.pxd

**Comment by jlaine** [comment]
Hm OK that didn't work for Windows. I still think it's the right way to go for POSIX platforms, as it avoids having to redefine these C functions. What does FFmpeg do on Windows?

File: av/container/core.pyx [hunk 1]
```diff
@@ -10,6 +11,9 @@ from av.container.output cimport OutputContainer
 from av.container.pyio cimport pyio_read, pyio_write, pyio_seek
 from av.format cimport build_container_format
 from av.utils cimport err_check, dict_to_avdict
+from posix.time cimport gettimeofday
+from av.container.core cimport cb_info
```
Comment (jlaine) [inline av/container/core.pyx]: Keep the imports sorted alpbatically, and regroup the imports from the same module

File: .gitignore [hunk 2]
```diff
@@ -3,6 +3,7 @@
 .DS_Store
 .nfs.*
 ._*
+.idea
```
Comment (jlaine) [inline .gitignore]: Same comment as for a previous commit: this doesn't have its place in this PR

File: av/container/core.pyx [hunk 3]
```diff
@@ -150,6 +171,12 @@ cdef class ContainerProxy(object):
             # Finish releasing the whole structure.
             lib.avformat_free_context(self.ptr)
 
+    def __set_callback_timeout__(self, timeout):
```
Comment (jlaine) [inline av/container/core.pyx]: These dunder methods look weird

File: av/utils.pxd [hunk 4]
```diff
@@ -1,4 +1,5 @@
 from libc.stdint cimport int64_t, uint8_t, uint64_t
+from posix.types cimport suseconds_t, time_t
```
Comment (jlaine) [inline av/utils.pxd]: Why do we need these imports?

File: av/container/core.pxd [hunk 5]
```diff
@@ -4,6 +4,11 @@ from av.container.streams cimport StreamContainer
 from av.dictionary cimport _Dictionary
 from av.format cimport ContainerFormat
 from av.stream cimport Stream
+from posix.time cimport timeval
+
+ctypedef struct cb_info:
+    int timeout
```
Comment (jlaine) [inline av/container/core.pxd]: Maybe store the timeout directly in milliseconds and likewise the start time in milliseconds, that way we don't drag in `timeval` here.

**Comment by rodrhern** [comment]
@jlaine I'll stop working on this PR, in favor of https://github.com/mikeboers/PyAV/pull/480
Thanks for your help

**Comment by jlaine** [comment]
OK let's move the discussion to #480

**Pull request closed without merging**